File: packages/runtime-web-three/src/sequences.ts
import type { ISequenceIr, ISequenceKeyframeIr, ISequencesIr, ISequenceTrackIr } from "@threenative/ir";

export interface ISequenceTraceInput {
  fixedDelta?: number;
  playByTick?: Record<number, readonly string[]>;
  skipByTick?: Record<number, readonly string[]>;
  stopByTick?: Record<number, readonly string[]>;
  ticks: number;
}

export interface ISequenceTraceObservation {
  entity?: string;
  kind: string;
  sequence: string;
  tick: number;
  time: number;
  track: string;
  value?: unknown;
}

export interface ISequenceTraceFrame {
  active: boolean;
  completed?: boolean;
  observations: ISequenceTraceObservation[];
  restoredCamera?: string;
  sequence: string;
  skipped?: boolean;
  stopped?: boolean;
  tick: number;
  time: number;
}

interface IActiveSequence {
  previousTime: number;
  time: number;
}

export function traceSequences(sequences: ISequencesIr, input: ISequenceTraceInput): ISequenceTraceFrame[] {
  const fixedDelta = input.fixedDelta ?? 0.5;
  const byId = new Map(sequences.sequences.map((sequence) => [sequence.id, sequence]));
  const active = new Map<string, IActiveSequence>();
  const trace: ISequenceTraceFrame[] = [];

  for (let tick = 0; tick < input.ticks; tick += 1) {
    for (const sequenceId of input.playByTick?.[tick] ?? []) {
      if (byId.has(sequenceId)) {
        active.set(sequenceId, { previousTime: 0, time: 0 });
      }
    }

    const forcedEnds = new Map<string, "skipped" | "stopped">();
    for (const sequenceId of input.skipByTick?.[tick] ?? []) {
      forcedEnds.set(sequenceId, "skipped");
    }
    for (const sequenceId of input.stopByTick?.[tick] ?? []) {
      forcedEnds.set(sequenceId, "stopped");
    }

    for (const [sequenceId, runtime] of [...active.entries()].sort(([left], [right]) => left.localeCompare(right))) {
      const sequence = byId.get(sequenceId);
      if (sequence === undefined) {
        active.delete(sequenceId);
        continue;
      }
      const forcedEnd = forcedEnds.get(sequenceId);
      const observations = forcedEnd === undefined ? sampleSequence(sequence, runtime.previousTime, runtime.time, tick) : [];
      const completed = forcedEnd === undefined && runtime.time >= sequence.duration;
      trace.push({
        active: forcedEnd === undefined && !completed,
        ...(completed ? { completed: true } : {}),
        observations,
        restoredCamera: sequenceCamera(sequence),
        sequence: sequence.id,
        ...(forcedEnd === "skipped" ? { skipped: true } : {}),
        ...(forcedEnd === "stopped" ? { stopped: true } : {}),
        tick,
        time: roundTime(runtime.time),
      });
      if (completed || forcedEnd !== undefined) {
        active.delete(sequenceId);
      } else {
        runtime.previousTime = runtime.time;
        runtime.time = Math.min(sequence.duration, runtime.time + fixedDelta);
      }
    }
  }

  return trace;
}

function sampleSequence(sequence: ISequenceIr, previousTime: number, time: number, tick: number): ISequenceTraceObservation[] {
  return sequence.tracks.flatMap((track) => {
    if (track.kind === "event" || track.kind === "audio" || track.kind === "ui") {
      return track.keyframes
        .filter((keyframe) => isTriggeredKey(keyframe, previousTime, time))
        .map((keyframe) => observation(sequence.id, track, tick, keyframe.time, keyframe.value));
    }
    return [observation(sequence.id, track, tick, time, sampleTrackValue(track, time))];
  });
}

function sampleTrackValue(track: ISequenceTrackIr, time: number): unknown {
  const keys = [...track.keyframes].sort((left, right) => left.time - right.time);
  if (keys.length === 0) {
    return undefined;
  }
  const first = keys[0];
  if (first === undefined) {
    return undefined;
  }
  const last = keys[keys.length - 1];
  if (last === undefined) {
    return undefined;
  }
  if (time <= first.time) {
    return first.value;
  }
  if (time >= last.time) {
    return last.value;
  }
  for (let index = 0; index < keys.length - 1; index += 1) {
    const left = keys[index];
    const right = keys[index + 1];
    if (left === undefined || right === undefined) {
      continue;
    }
    if (time >= left.time && time <= right.time) {
      if (right.easing === "step" || left.easing === "step") {
        return left.value;
      }
      return lerpValue(left.value, right.value, (time - left.time) / (right.time - left.time));
    }
  }
  return last.value;
}

function lerpValue(left: unknown, right: unknown, t: number): unknown {
  if (typeof left === "number" && typeof right === "number") {
    return roundTime(left + (right - left) * t);
  }
  if (Array.isArray(left) && Array.isArray(right) && left.length === right.length) {
    return left.map((item, index) => lerpValue(item, right[index], t));
  }
  if (isRecord(left) && isRecord(right)) {
    return Object.fromEntries(Object.keys(left).sort().map((key) => [key, lerpValue(left[key], right[key], t)]));
  }
  return t < 1 ? left : right;
}

function observation(sequence: string, track: ISequenceTrackIr, tick: number, time: number, value: unknown): ISequenceTraceObservation {
  return {
    ...(track.entity === undefined ? {} : { entity: track.entity }),
    kind: track.kind,
    sequence,
    tick,
    time: roundTime(time),
    track: track.id,
    ...(value === undefined ? {} : { value }),
  };
}

function sequenceCamera(sequence: ISequenceIr): string | undefined {
  return sequence.tracks.find((track) => track.kind === "cameraPose")?.entity;
}

function isTriggeredKey(keyframe: ISequenceKeyframeIr, previousTime: number, time: number): boolean {
  return (time === 0 && keyframe.time === 0) || (keyframe.time > previousTime && keyframe.time <= time);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function roundTime(value: number): number {
  return Number(value.toFixed(6));
}

File: runtime-bevy/crates/threenative_runtime/src/sequences.rs
use std::collections::HashMap;

use serde::Serialize;
use threenative_loader::{SequenceIr, SequenceKeyframeIr, SequenceTrackIr, SequencesIr};

#[derive(Clone, Debug, Default)]
pub struct NativeSequenceTraceInput {
    pub fixed_delta: f32,
    pub play_by_tick: HashMap<u32, Vec<String>>,
    pub skip_by_tick: HashMap<u32, Vec<String>>,
    pub stop_by_tick: HashMap<u32, Vec<String>>,
    pub ticks: u32,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeSequenceTraceFrame {
    pub active: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub completed: Option<bool>,
    pub observations: Vec<NativeSequenceTraceObservation>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub restored_camera: Option<String>,
    pub sequence: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub skipped: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stopped: Option<bool>,
    pub tick: u32,
    pub time: f32,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct NativeSequenceTraceObservation {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub entity: Option<String>,
    pub kind: String,
    pub sequence: String,
    pub tick: u32,
    pub time: f32,
    pub track: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub value: Option<serde_json::Value>,
}

#[derive(Clone, Debug)]
struct ActiveSequence {
    previous_time: f32,
    time: f32,
}

pub fn trace_sequences(
    sequences: &SequencesIr,
    input: NativeSequenceTraceInput,
) -> Vec<NativeSequenceTraceFrame> {
    let fixed_delta = if input.fixed_delta > 0.0 {
        input.fixed_delta
    } else {
        0.5
    };
    let by_id: HashMap<&str, &SequenceIr> = sequences
        .sequences
        .iter()
        .map(|sequence| (sequence.id.as_str(), sequence))
        .collect();
    let mut active: HashMap<String, ActiveSequence> = HashMap::new();
    let mut trace = Vec::new();

    for tick in 0..input.ticks {
        if let Some(play) = input.play_by_tick.get(&tick) {
            for sequence_id in play {
                if by_id.contains_key(sequence_id.as_str()) {
                    active.insert(
                        sequence_id.clone(),
                        ActiveSequence {
                            previous_time: 0.0,
                            time: 0.0,
                        },
                    );
                }
            }
        }

        let mut forced_ends: HashMap<String, &str> = HashMap::new();
        if let Some(skip) = input.skip_by_tick.get(&tick) {
            for sequence_id in skip {
                forced_ends.insert(sequence_id.clone(), "skipped");
            }
        }
        if let Some(stop) = input.stop_by_tick.get(&tick) {
            for sequence_id in stop {
                forced_ends.insert(sequence_id.clone(), "stopped");
            }
        }

        let mut active_ids: Vec<_> = active.keys().cloned().collect();
        active_ids.sort();
        for sequence_id in active_ids {
            let Some(sequence) = by_id.get(sequence_id.as_str()) else {
                active.remove(&sequence_id);
                continue;
            };
            let Some(runtime) = active.get_mut(&sequence_id) else {
                continue;
            };
            let forced_end = forced_ends.get(&sequence_id).copied();
            let observations = if forced_end.is_none() {
                sample_sequence(sequence, runtime.previous_time, runtime.time, tick)
            } else {
                Vec::new()
            };
            let completed = forced_end.is_none() && runtime.time >= sequence.duration;
            trace.push(NativeSequenceTraceFrame {
                active: forced_end.is_none() && !completed,
                completed: completed.then_some(true),
                observations,
                restored_camera: sequence_camera(sequence),
                sequence: sequence.id.clone(),
                skipped: (forced_end == Some("skipped")).then_some(true),
                stopped: (forced_end == Some("stopped")).then_some(true),
                tick,
                time: round_time(runtime.time),
            });
            if completed || forced_end.is_some() {
                active.remove(&sequence_id);
            } else {
                runtime.previous_time = runtime.time;
                runtime.time = sequence.duration.min(runtime.time + fixed_delta);
            }
        }
    }

    trace
}

fn sample_sequence(
    sequence: &SequenceIr,
    previous_time: f32,
    time: f32,
    tick: u32,
) -> Vec<NativeSequenceTraceObservation> {
    let mut observations = Vec::new();
    for track in &sequence.tracks {
        if matches!(track.kind.as_str(), "event" | "audio" | "ui") {
            for keyframe in track
                .keyframes
                .iter()
                .filter(|keyframe| is_triggered_key(keyframe, previous_time, time))
            {
                observations.push(observation(
                    &sequence.id,
                    track,
                    tick,
                    keyframe.time,
                    keyframe.value.clone(),
                ));
            }
        } else {
            observations.push(observation(
                &sequence.id,
                track,
                tick,
                time,
                sample_track_value(track, time),
            ));
        }
    }
    observations
}

fn sample_track_value(track: &SequenceTrackIr, time: f32) -> Option<serde_json::Value> {
    let mut keys = track.keyframes.clone();
    keys.sort_by(|left, right| left.time.total_cmp(&right.time));
    let first = keys.first()?;
    let last = keys.last()?;
    if time <= first.time {
        return first.value.clone();
    }
    if time >= last.time {
        return last.value.clone();
    }
    for window in keys.windows(2) {
        let left = &window[0];
        let right = &window[1];
        if time >= left.time && time <= right.time {
            if right.easing.as_deref() == Some("step") || left.easing.as_deref() == Some("step") {
                return left.value.clone();
            }
            let t = (time - left.time) / (right.time - left.time);
            return lerp_value(left.value.as_ref()?, right.value.as_ref()?, t);
        }
    }
    last.value.clone()
}

fn lerp_value(
    left: &serde_json::Value,
    right: &serde_json::Value,
    t: f32,
) -> Option<serde_json::Value> {
    match (left, right) {
        (serde_json::Value::Number(left), serde_json::Value::Number(right)) => {
            let left = left.as_f64()?;
            let right = right.as_f64()?;
            Some(serde_json::Value::from(round_time(
                (left + (right - left) * f64::from(t)) as f32,
            )))
        }
        (serde_json::Value::Array(left), serde_json::Value::Array(right))
            if left.len() == right.len() =>
        {
            let values: Option<Vec<_>> = left
                .iter()
                .zip(right.iter())
                .map(|(left, right)| lerp_value(left, right, t))
                .collect();
            values.map(serde_json::Value::Array)
        }
        (serde_json::Value::Object(left), serde_json::Value::Object(right)) => {
            let mut keys: Vec<_> = left.keys().collect();
            keys.sort();
            let mut value = serde_json::Map::new();
            for key in keys {
                value.insert(key.clone(), lerp_value(left.get(key)?, right.get(key)?, t)?);
            }
            Some(serde_json::Value::Object(value))
        }
        _ => Some(if t < 1.0 { left.clone() } else { right.clone() }),
    }
}

fn observation(
    sequence: &str,
    track: &SequenceTrackIr,
    tick: u32,
    time: f32,
    value: Option<serde_json::Value>,
) -> NativeSequenceTraceObservation {
    NativeSequenceTraceObservation {
        entity: track.entity.clone(),
        kind: track.kind.clone(),
        sequence: sequence.to_owned(),
        tick,
        time: round_time(time),
        track: track.id.clone(),
        value,
    }
}

fn sequence_camera(sequence: &SequenceIr) -> Option<String> {
    sequence
        .tracks
        .iter()
        .find(|track| track.kind == "cameraPose")
        .and_then(|track| track.entity.clone())
}

fn is_triggered_key(keyframe: &&SequenceKeyframeIr, previous_time: f32, time: f32) -> bool {
    (time == 0.0 && keyframe.time == 0.0)
        || (keyframe.time > previous_time && keyframe.time <= time)
}

fn round_time(value: f32) -> f32 {
    (value * 1_000_000.0).round() / 1_000_000.0
}
